[macOS] "text/uri-list" is always empty when dragging and dropping from the unified field in Safari

whsieh opened the pull request:
#### 15a84a83d45966d96d4bc5f5a54a3e47e490cd14
<pre>
[macOS] &quot;text/uri-list&quot; is always empty when dragging and dropping from the unified field in Safari
<a href="https://bugs.webkit.org/show_bug.cgi?id=242143">https://bugs.webkit.org/show_bug.cgi?id=242143</a>
rdar://96074318

Reviewed by Devin Rousso.

When dragging from the unified search field on macOS, Safari writes a promised `.webloc` file to the
pasteboard. This causes us to go down the `legacyFilesPromisePasteboardType()` handling codepath in
`WebViewImpl::performDragOperation`, which asynchronously receives promised files from the drag
pasteboard, creates sandbox extensions for these promised files, and then calls into
`performDragOperation()` with these files and extensions.

However, this codepath is missing a call to `grantAccessToCurrentPasteboardData`, which means that
the access level for the drag pasteboard that&apos;s granted to the web page when dropping promised files
is still `PasteboardAccessType::Types` rather than `PasteboardAccessType::TypesAndData`. (Somewhat
surprisingly), this doesn&apos;t have much effect when dragging and dropping files, since the actual data
in each of dropped files is accessible through file URLs on the `DragData` that&apos;s already in the web
process, and we generally only end up asking the platform pasteboard for the list of types (in order
to expose &quot;Files&quot; as one of the `DataTransfer` types).

However, this means that in the case where we&apos;re dropping a `.webloc` file, we end up exposing the
`&quot;text/uri-list&quot;` type on the DataTransfer (since the web process is able to read the list of
pasteboard types), but we&apos;re unable to actually grab the URL, since the UI process hasn&apos;t granted
access to `TypesAndData`.

To fix this, we simply move the check down into `WebPageProxy::performDragOperation`, instead of
requiring each individual call site to `grantAccessToCurrentPasteboardData()`. This also allows us
to remove an explicit call to `grantAccessToCurrentPasteboardData()` in iOS-specific drop handling
code in `WKContentViewInteraction.mm`; see below for more details.

Test: DragAndDropTests.ReadURLWhenDroppingPromisedWebLoc

* Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
(WebKit::WebViewImpl::performDragOperation):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::performDragOperation):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:

Consolidate all calls to `grantAccessToCurrentPasteboardData` during drop into `WebPageProxy`.

(-[WKContentView dropInteraction:performDrop:]):
* Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm:
* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
* Tools/TestWebKitAPI/mac/DragAndDropSimulatorMac.mm:
(-[DragAndDropSimulator writePromisedWebLoc:]):
* Tools/TestWebKitAPI/mac/TestFilePromiseReceiver.mm:
(writeToWebLoc):
(-[TestFilePromiseReceiver receivePromisedFilesAtDestination:options:operationQueue:reader:]):

Add support for `DragAndDropSimulator` to write a `.webloc` file with a URL when dragging; this
simulates dragging from the unified field in Safari.

Canonical link: <a href="https://commits.webkit.org/251996@main">https://commits.webkit.org/251996@main</a>
</pre>

dcrousso commented on Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
```diff
@@ -4386,7 +4386,6 @@ static NSDragOperation kit(std::optional<WebCore::DragOperation> dragOperation)
     }
 
     String draggingPasteboardName = draggingInfo.draggingPasteboard.name;
-    m_page->grantAccessToCurrentPasteboardData(draggingPasteboardName);
```
Out of curiosity, would this bug have been fixed by copying this line to before `m_page->performDragOperation` above?

whsieh committed:
[macOS] "text/uri-list" is always empty when dragging and dropping from the unified field in Safari
https://bugs.webkit.org/show_bug.cgi?id=242143
rdar://96074318

Reviewed by Devin Rousso.

When dragging from the unified search field on macOS, Safari writes a promised `.webloc` file to the
pasteboard. This causes us to go down the `legacyFilesPromisePasteboardType()` handling codepath in
`WebViewImpl::performDragOperation`, which asynchronously receives promised files from the drag
pasteboard, creates sandbox extensions for these promised files, and then calls into
`performDragOperation()` with these files and extensions.

However, this codepath is missing a call to `grantAccessToCurrentPasteboardData`, which means that
the access level for the drag pasteboard that's granted to the web page when dropping promised files
is still `PasteboardAccessType::Types` rather than `PasteboardAccessType::TypesAndData`. (Somewhat
surprisingly), this doesn't have much effect when dragging and dropping files, since the actual data
in each of dropped files is accessible through file URLs on the `DragData` that's already in the web
process, and we generally only end up asking the platform pasteboard for the list of types (in order
to expose "Files" as one of the `DataTransfer` types).

However, this means that in the case where we're dropping a `.webloc` file, we end up exposing the
`"text/uri-list"` type on the DataTransfer (since the web process is able to read the list of
pasteboard types), but we're unable to actually grab the URL, since the UI process hasn't granted
access to `TypesAndData`.

To fix this, we simply move the check down into `WebPageProxy::performDragOperation`, instead of
requiring each individual call site to `grantAccessToCurrentPasteboardData()`. This also allows us
to remove an explicit call to `grantAccessToCurrentPasteboardData()` in iOS-specific drop handling
code in `WKContentViewInteraction.mm`; see below for more details.

Test: DragAndDropTests.ReadURLWhenDroppingPromisedWebLoc

* Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
(WebKit::WebViewImpl::performDragOperation):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::performDragOperation):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:

Consolidate all calls to `grantAccessToCurrentPasteboardData` during drop into `WebPageProxy`.

(-[WKContentView dropInteraction:performDrop:]):
* Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm:
* Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h:
* Tools/TestWebKitAPI/mac/DragAndDropSimulatorMac.mm:
(-[DragAndDropSimulator writePromisedWebLoc:]):
* Tools/TestWebKitAPI/mac/TestFilePromiseReceiver.mm:
(writeToWebLoc):
(-[TestFilePromiseReceiver receivePromisedFilesAtDestination:options:operationQueue:reader:]):

Add support for `DragAndDropSimulator` to write a `.webloc` file with a URL when dragging; this
simulates dragging from the unified field in Safari.

Canonical link: https://commits.webkit.org/251996@main
